src/clj/fluree/db/vector/index_graph.cljc

```diff
@@ -0,0 +1,171 @@
+(ns fluree.db.vector.index-graph
+  (:require [clojure.string :as str]
+            [fluree.db.constants :as const]
+            [fluree.db.flake :as flake]
+            [fluree.db.json-ld.iri :as iri]
+            [fluree.db.query.exec.where :as where]
+            [clojure.core.async :as async]
+            [fluree.db.util.async :refer [go-try <?]]
+            [fluree.db.query.range :as query-range]
+            [fluree.db.vector.scoring :as vector.score]
+            [fluree.db.util.core :refer [try* catch*]]
+            [fluree.db.util.log :as log]))
+
+(def iri-compare (str iri/f-idx-flatrank-ns "compare"))
+(def iri-property (str iri/f-idx-flatrank-ns "property"))
+(def iri-limit (str iri/f-idx-flatrank-ns "limit"))
+(def iri-id (str iri/f-idx-flatrank-ns "id"))
+(def iri-score (str iri/f-idx-flatrank-ns "score"))
+(def iri-vector (str iri/f-idx-flatrank-ns "vector"))
+(def iri-xsd-float "http://www.w3.org/2001/XMLSchema#float")
+
+(def metric-re (re-pattern (str iri/f-idx-flatrank-ns "(.*)")))
+
+(def result-sort (fn [a b] (compare (get a :score) (get b :score))))
+(def reverse-result-sort (fn [a b] (compare (get b :score) (get a :score))))
+
+(def metrics
+  {:dotproduct {::score-fn vector.score/dotproduct
+                ::sort-fn  reverse-result-sort}
+   :cosine     {::score-fn vector.score/cosine-similarity
+                ::sort-fn  reverse-result-sort}
+   :distance   {::score-fn vector.score/euclidian-distance
+                ::sort-fn  result-sort}})
+
+(defn- prop-iri
+  [triple]
+  (-> triple (nth 1) ::where/iri))
+
+(defn- obj-val
+  [triple solution]
+  (let [o (nth triple 2)]
+    (or (where/get-value o)
+        (->> (where/get-variable o)
+             (get solution)
+             (where/get-value)))))
+
+(defn- obj-var
+  [triple]
+  (-> triple (nth 2) ::where/var))
+
+(defn- obj-iri
+  [triple]
+  (-> triple (nth 2) ::where/iri))
+
+(defn extract-metric
+  "Takes the graph alias as a string and extracts the metric name from the
+  end of the IRI"
+  [graph-alias]
+  (some-> (re-find metric-re graph-alias)
+          second
+          str/lower-case
+          keyword))
+
+(defn parse-search-graph
+  [graph-alias solution graph-triples]
+  (try*
+    (let [metric (extract-metric graph-alias)]
+      (reduce
+       (fn [acc triple]
+         (cond
+           (= iri-compare (prop-iri triple))
+           (assoc acc ::compare (obj-val triple solution))
+
+           (= iri-property (prop-iri triple))
+           (assoc acc ::property (obj-iri triple))
+
+           (= iri-limit (prop-iri triple))
+           (assoc acc ::limit (obj-val triple solution))
+
+           (= iri-id (prop-iri triple))
+           (assoc-in acc [::result ::id] (obj-var triple))
+
+           (= iri-score (prop-iri triple))
+           (assoc-in acc [::result ::score] (obj-var triple))
+
+           (= iri-vector (prop-iri triple))
```

Comment:
Might be worth pulling the result of `(prop-iri triple)` into a `let` to avoid re-evaluating it each time.

Reply:
Yes good point. This is only hit at parse time, so won't have a measurable impact on speed - but should have done this regadless.

Reply:
Addressed in c749929213c37034e4d187e61ed969049197a5d0